### PR TITLE
Fix SSA being silently disabled by lack of OpenAPI config

### PR DESCRIPTION
**Pull request by alexzielenski**
#### What type of PR is this?

/kind bug
/kind cleanup

#### What this PR does / why we need it:

We've previously had the condition where, despite SSA being enabled, it would be silently disabled if the apiserver library user did not provide OpenAPI config and models for SSA to create a type converter. Since SSA is enabled by default and locked to default, this is surprising. In fact, despite the top-level kube-apiserver making proper use of SSA by providing these models, many unit/integration tests are not testing the SSA path since they in some way do not provide the OpenAPI config/definitions.

This PR:

1.) Makes it an error to not provide openapi definitions for each API resource installed during startup
2.) Fixes up tests to now provide proper OpenAPI schemas for their test servers (previously they would be testing with SSA effectively disabled)
3.) Cleans up conditional statements about SSA being enabled that are no longer required

#### Which issue(s) this PR fixes:

Fixes #

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs
NONE
```

 Related: https://github.com/kubernetes/kubernetes/pull/114439

/sig api-machinery
/assign @apelisse

### CODE REVIEW

File: staging/src/k8s.io/apiserver/pkg/server/genericapiserver.go [hunk 1]
```diff
@@ -883,7 +880,9 @@ func NewDefaultAPIGroupInfo(group string, scheme *runtime.Scheme, parameterCodec
 // getOpenAPIModels is a private method for getting the OpenAPI models
 func (s *GenericAPIServer) getOpenAPIModels(apiPrefix string, apiGroupInfos ...*APIGroupInfo) (*spec.Swagger, error) {
 	if s.openAPIConfig == nil {
-		return nil, nil
+		// SSA is GA (enabled by default) and requires OpenAPI config to be set
```
Comment (deads2k) [inline staging/src/k8s.io/apiserver/pkg/server/genericapiserver.go:883]: I agree.  In my opinion, to be a kube-compatible API at this point, it must support SSA. Aggregated apiservers serving their APIs must participate to continue updating past this point.

File: staging/src/k8s.io/apiserver/pkg/server/genericapiserver.go [hunk 2]
```diff
@@ -667,6 +668,11 @@ func (s preparedGenericAPIServer) NonBlockingRun(stopCh <-chan struct{}, shutdow
 
 // installAPIResources is a private method for installing the REST storage backing each api groupversionresource
 func (s *GenericAPIServer) installAPIResources(apiPrefix string, apiGroupInfo *APIGroupInfo, openAPIModels *spec.Swagger) error {
+	typeConverter, err := fieldmanager.NewTypeConverter(openAPIModels, false)
```
Comment (deads2k) [inline staging/src/k8s.io/apiserver/pkg/server/genericapiserver.go:671]: We used to have an instance per groupversion.  I'm not familiar enough with the fieldmanager to know if that matters offhand.

cc @apelisse

**Comment by deads2k** [comment]
Adding the requirement for openapi in order to add an API makes sense at this point so that clients can have predictable interactions.

/approve
/assign @apelisse

File: staging/src/k8s.io/apiserver/pkg/server/genericapiserver.go (hunk 2 above)
Comment (apelisse) [reply staging/src/k8s.io/apiserver/pkg/server/genericapiserver.go:671]: We don't need multiple copies of these, they have the same input parameters and so are all going to be exact copies anyway. They are expensive to build and keep a lot of allocated memory in-use. They are considered immutable (except for some internal caching). And they all have the schema for all the internal types anyway. We're making their construction A LOT cheaper in #114439, but it's still much better to just do it once.

**Comment by apelisse** [comment]
What's blocking this by the way?

**Comment by alexzielenski** [comment]
https://github.com/kubernetes/kubernetes/pull/115324 changes to use openAPIV3Config instead of openAPIConfig, so I did not want to make it an error to not have `openAPIConfig` right before changing the requirement to being `openAPIV3Config`

**Review by apelisse — COMMENTED** [review]
As mentioned offline, there are many places where we check if typeconverter is set, maybe these should be updated too

**Comment by apelisse** [comment]
I'm excited to see that thread revived.

File: staging/src/k8s.io/apiserver/pkg/server/genericapiserver.go [hunk 3]
```diff
@@ -977,7 +969,13 @@ func (s *GenericAPIServer) getOpenAPIModels(apiPrefix string, apiGroupInfos ...*
...
 		apiGroupInfo.StaticOpenAPISpec = openAPISpec
 	}
-	return openAPISpec, nil
+
+	typeConverter, err := managedfields.NewTypeConverter(openAPISpec, false)
+	if err != nil {
+		return nil, err
+	}
+
+	return typeConverter, nil
```
Comment (apelisse) [inline staging/src/k8s.io/apiserver/pkg/server/genericapiserver.go:978]: Could be `return managedfields.NewTypeConverter(openAPISpec, false)`

File: staging/src/k8s.io/apiserver/pkg/endpoints/handlers/patch.go [hunk 4]
```diff
@@ -344,6 +344,11 @@ func (p *jsonPatcher) applyPatchToCurrentObject(requestContext context.Context,
 		}
 	}
 
+	if p.options == nil {
+		// Provide a more informative error for the crash that would
+		// happen on the next line
+		panic("PatchOptions required but not provided")
+	}
```
Comment (apelisse) [inline staging/src/k8s.io/apiserver/pkg/endpoints/handlers/patch.go:351]: Have you seen that panic? Clearly it shouldn't happen. I'm guessing that the added options below?

**Review by apelisse — COMMENTED** [review]
Minor nits, thanks!

**Comment by RainbowMango** [comment]
Hey PR author and approvers, @alexzielenski @deads2k @apelisse
Do you think this patch should be backported to release branches?
I'm asking because [this issue hasn't been fixed](https://github.com/kubernetes/kubernetes/blob/6d31f4b31ba1294c4844ec94ea2f67c5f8bf42f4/staging/src/k8s.io/apiserver/pkg/server/genericapiserver.go#L962) on release-1.27 and might block people from bumping their K8S dependencies.

**Comment by apelisse** [comment]
@RainbowMango I don't think it's likely we will back-port that change, but I also don't understand what problem you have that would require this to be back-ported for? Can you elaborate?

**Comment by RainbowMango** [comment]
@apelisse Thanks, but please ignore me now. We had some block issues(OpenAPI related) during [the update of Kubernetes dependencies to v1.27.3](https://github.com/karmada-io/karmada/pull/3730), but now it has been solved.

**Pull request merged**